### ISSUE TITLE
Generate compile_commands.json file needed by clangd VS Code plugin (Bugzilla Bug 2850)

### ISSUE THREAD

**Issue by tianocore-issues** [tianocore/edk2#8547]
This issue was created automatically with [bugzilla2github](https://github.com/makubacki/bugzilla2github/tree/bz2gh_tianocore)
# Bugzilla Bug 2850
Date: 2020-07-14T02:07:43+00:00
From: @niruiyu
To:   Guillermo A Palomino Sosa &lt;<guillermo.a.palomino.sosa>&gt;
CC:   @BobCF, daniel.schaefer, @lgao4, guillermo.a.palomino.sosa, jian.j.wang, nobody, @YuweiChen1110

Last updated: 2023-03-13T16:35:39+00:00

**Comment by tianocore-issues** [comment]
## Comment 13096
Date: 2020-07-14 02:07:43 +0000
From: @niruiyu

- Industry Specification: ---
- Release Observed: EDK II Master
- Releases to Fix: EDK II Master
- Target OS: ---
- Bugzilla Assignee(s): Guillermo A Palomino Sosa &lt;<guillermo.a.palomino.sosa>&gt;

In order for clangd to understand the source code, a compile_commands.json file needs to supplied to clangd.

Refer to http://clang.llvm.org/docs/JSONCompilationDatabase.html for detailed information on how the file looks like.

Basically the file contains the information of how each source file is compiled.

In my opinion, the information is similar to what a top-level make file contains.

CMake project can generate such file by
 setting -DCMAKE_EXPORT_COMPILE_COMMANDS=1.

**Comment by tianocore-issues** [comment]
## Comment 13204
Date: 2020-07-28 22:05:46 +0000
From: nobody &lt;<nobody>&gt;

Ray: can you provide more background for this request?

**Comment by tianocore-issues** [comment]
## Comment 13765
Date: 2020-09-23 02:34:23 +0000
From: @niruiyu

I am using Visual Studio Code and found its default source code navigation doesn&apos;t work very well. There is a clangd extension that supports code navigation as well but it needs the compile_commands.json. Actually this json file is very similar to the top level make file once we discussed in tool domain.

**Comment by tianocore-issues** [comment]
## Comment 13846
Date: 2020-09-29 21:48:48 +0000
From: @lgao4

Ray: would you contribute the proposal and change for this request?

**Comment by tianocore-issues** [comment]
## Comment 15065
Date: 2021-02-12 03:49:15 +0000
From: Daniel Schaefer &lt;<daniel.schaefer>&gt;

Someone has made a patch for EDK2 to support this: https://github.com/makaleks/edk2-tools/tree/master/compilation_database_patch
I tried it on the edk2-stable202008 tag but it doesn&apos;t work anymore. My build crashes without a useful error message. I haven&apos;t tried to debug this.
Since it&apos;s a very simple patch, I think it&apos;s not a bad idea to try to make this work again.

What I did instead was to use the bear [1] tool and run the EDK2 build using that. It intercepts all the compiler calls and records them. That works fine.

[1] https://github.com/rizsotto/Bear

**Comment by tianocore-issues** [comment]
## Comment 20997
Date: 2023-02-07 12:26:04 +0000
From: Guillermo A Palomino Sosa &lt;<guillermo.a.palomino.sosa>&gt;

I have sent a patch that extends Build report functions to create the build database +some other useful files:
https://edk2.groups.io/g/devel/message/99716

This adds the -Y REPORT_INFO and will create following files in the build folder:

&lt;Build&gt;/&lt;BuildTarget&gt;/&lt;ToolChain&gt;/CompileInfo

* compile_commands.json - Compilation Database. To be used by IDE&apos;s
  to enable advance features
* cscope.files - List of files used in compilation. Used by Cscope to parse
  C code and provide browse functionality.
* module_report.json - Module data form buildReport in Json format.

**Comment by tianocore-issues** [comment]
## Comment 20998
Date: 2023-02-07 12:27:19 +0000
From: Guillermo A Palomino Sosa &lt;<guillermo.a.palomino.sosa>&gt;

Created attachment 1320
Extends build report functionality with -Y REPORT_INFO

Attachment: [BZ-1320-0001-BaseTools-Generate-compile-information-in-build-repo.patch](https://github.com/tianocore/user-attachments/raw/refs/heads/main/tianocore/edk2/BZ-1320-0001-BaseTools-Generate-compile-information-in-build-repo.patch)

**Comment by tianocore-issues** [comment]
## Comment 21059
Date: 2023-02-14 21:59:02 +0000
From: @lgao4

Include Bob. 

For new feature, it should be added into BaseTools Python project first (https://github.com/tianocore/edk2-basetools).

**Comment by tianocore-issues** [comment]
## Comment 21123
Date: 2023-02-28 21:34:42 +0000
From: @lgao4

Guillermo has sent the patch for review.

**Comment by tianocore-issues** [comment]
## Comment 21182
Date: 2023-03-13 16:35:39 +0000
From: Guillermo A Palomino Sosa &lt;<guillermo.a.palomino.sosa>&gt;

Code has been submitted:
 https://github.com/tianocore/edk2/commit/4ad7ea9c842289aed6ed519f9fce33cf37cfc2a9
